clippy_lints/src/erasing_op.rs

```diff
@@ -37,22 +39,32 @@ impl<'tcx> LateLintPass<'tcx> for ErasingOp {
         if let ExprKind::Binary(ref cmp, left, right) = e.kind {
             match cmp.node {
                 BinOpKind::Mul | BinOpKind::BitAnd => {
-                    check(cx, left, e.span);
-                    check(cx, right, e.span);
+                    check(cx, left, e);
+                    check(cx, right, e);
                 },
-                BinOpKind::Div => check(cx, left, e.span),
+                BinOpKind::Div => check(cx, left, e),
                 _ => (),
             }
         }
     }
 }
 
-fn check(cx: &LateContext<'_>, e: &Expr<'_>, span: Span) {
-    if constant_simple(cx, cx.typeck_results(), e) == Some(Constant::Int(0)) {
+fn different_types(typeck_results: &TypeckResults<'tcx>, left: &'tcx Expr<'_>, right: &'tcx Expr<'_>) -> bool {
```

Comment:
NIT: Here I would try to give the two arguments a better name to indicate what they are. At first when reading this I suspected that they are the left and right expression of the multiplication, while they are actually the input and output. Could you maybe update this: 
```suggestion
fn different_types(typeck_results: &TypeckResults<'tcx>, input: &'tcx Expr<'_>, output: &'tcx Expr<'_>) -> bool {
```